File: scripts/download_datalab_cache.py
from __future__ import annotations

import json
import os
import time
from pathlib import Path

import httpx


BASE_URL = "https://models.datalab.to"
CHECKPOINTS = (
    "layout/2025_09_23",
    "text_detection/2025_05_07",
    "text_recognition/2025_09_23",
    "table_recognition/2025_02_18",
    "ocr_error_detection/2025_02_18",
)
CHUNK_SIZE = 4 * 1024 * 1024


def model_root() -> Path:
    cache_home = Path(
        os.environ.get(
            "XDG_CACHE_HOME",
            Path(__file__).resolve().parents[1] / "runs" / "tools" / "cache",
        )
    )
    return cache_home / "datalab" / "models"


def content_length(url: str) -> int | None:
    with httpx.Client(
        follow_redirects=True,
        headers={"User-Agent": "Mozilla/5.0"},
        timeout=httpx.Timeout(60, connect=20, read=60),
    ) as client:
        response = client.head(url)
        response.raise_for_status()
        value = response.headers.get("Content-Length")
    return int(value) if value else None


def download(url: str, destination: Path, expected_size: int | None) -> None:
    if destination.is_file() and (
        expected_size is None or destination.stat().st_size == expected_size
    ):
        print(f"skip {destination.relative_to(model_root())}")
        return

    destination.parent.mkdir(parents=True, exist_ok=True)
    partial = destination.with_suffix(destination.suffix + ".part")
    for attempt in range(1, 11):
        downloaded = partial.stat().st_size if partial.exists() else 0
        headers = {
            "User-Agent": "Mozilla/5.0",
            "Cache-Control": "no-cache",
            "Pragma": "no-cache",
        }
        if downloaded:
            headers["Range"] = f"bytes={downloaded}-"

        try:
            timeout = httpx.Timeout(60, connect=20, read=30)
            separator = "&" if "?" in url else "?"
            request_url = f"{url}{separator}cachebust={time.time_ns()}-{attempt}"
            with httpx.stream(
                "GET",
                request_url,
                headers=headers,
                follow_redirects=True,
                timeout=timeout,
            ) as response:
                response.raise_for_status()
                if downloaded and response.status_code != 206:
                    partial.unlink(missing_ok=True)
                    downloaded = 0
                with partial.open("ab" if downloaded else "wb") as handle:
                    for chunk in response.iter_bytes(chunk_size=CHUNK_SIZE):
                        if chunk:
                            handle.write(chunk)
            break
        except httpx.HTTPError as exc:
            if attempt == 10:
                raise
            size = partial.stat().st_size if partial.exists() else 0
            print(
                f"retry {attempt}/10 {destination.relative_to(model_root())} "
                f"after {type(exc).__name__}, partial={size}",
                flush=True,
            )
            time.sleep(2 * attempt)

    size = partial.stat().st_size
    if expected_size is not None and size != expected_size:
        raise RuntimeError(f"bad size for {destination}: {size} != {expected_size}")
    partial.replace(destination)
    print(f"downloaded {destination.relative_to(model_root())} {size}")


def checkpoint_files(root: Path, checkpoint: str) -> list[str]:
    manifest_path = root / checkpoint / "manifest.json"
    if not manifest_path.is_file():
        raise FileNotFoundError(f"missing manifest: {manifest_path}")
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    files = manifest.get("files")
    if not isinstance(files, list):
        raise RuntimeError(f"invalid manifest files list: {manifest_path}")
    return [str(file) for file in files]


def main() -> int:
    root = model_root()
    root.mkdir(parents=True, exist_ok=True)
    for checkpoint in CHECKPOINTS:
        files = checkpoint_files(root, checkpoint)
        for file_name in files:
            destination = root / checkpoint / file_name
            if destination.is_file():
                print(f"skip {destination.relative_to(root)}", flush=True)
                continue
            if file_name == "model.safetensors":
                raise FileNotFoundError(
                    f"missing local model weight; automatic weight download is disabled: {destination}"
                )
            url = f"{BASE_URL}/{checkpoint}/{file_name}"
            download(url, destination, expected_size=None)
            time.sleep(0.1)
    print("datalab model cache complete")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/download_datalab_cache.mjs
import fs from "node:fs";
import { mkdir, rename, stat, unlink } from "node:fs/promises";
import path from "node:path";

const BASE_URL = "https://models.datalab.to";
const CHECKPOINTS = [
  "layout/2025_09_23",
  "text_detection/2025_05_07",
  "text_recognition/2025_09_23",
  "table_recognition/2025_02_18",
  "ocr_error_detection/2025_02_18",
];
const CHUNK_SIZE = 24 * 1024;
const PROGRESS_STEP = 100 * 1024 * 1024;

const cacheHome =
  process.env.XDG_CACHE_HOME ||
  path.resolve("runs", "tools", "cache");
const modelRoot = path.join(cacheHome, "datalab", "models");
const includeWeights = process.argv.includes("--include-weights");

async function fileSize(file) {
  try {
    return (await stat(file)).size;
  } catch {
    return null;
  }
}

async function expectedSize(url) {
  for (let attempt = 1; attempt <= 10; attempt += 1) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 15_000);
    try {
      const separator = url.includes("?") ? "&" : "?";
      const response = await fetch(`${url}${separator}head=${Date.now()}_${attempt}`, {
        method: "HEAD",
        headers: { "user-agent": "Mozilla/5.0", "cache-control": "no-cache" },
        signal: controller.signal,
      });
      if (!response.ok) {
        throw new Error(`HEAD ${url}: ${response.status}`);
      }
      const length = response.headers.get("content-length");
      return length ? Number(length) : null;
    } catch (error) {
      if (attempt === 10) throw error;
    } finally {
      clearTimeout(timer);
    }
  }
}

async function download(url, destination, expected) {
  const current = await fileSize(destination);
  if (current !== null && (expected === null || current === expected)) {
    console.log(`skip ${path.relative(modelRoot, destination)}`);
    return;
  }

  await mkdir(path.dirname(destination), { recursive: true });
  const partial = `${destination}.part`;
  let offset = (await fileSize(partial)) || 0;

  if (expected === null) {
    throw new Error(`missing content-length for ${url}`);
  }

  let lastProgress = offset;
  while (offset < expected) {
    const end = Math.min(offset + CHUNK_SIZE - 1, expected - 1);
    let wrote = false;
    for (let attempt = 1; attempt <= 20; attempt += 1) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), 15_000);
      try {
        const separator = url.includes("?") ? "&" : "?";
        const response = await fetch(`${url}${separator}download=${Date.now()}_${attempt}`, {
          headers: {
            "user-agent": "Mozilla/5.0",
            "cache-control": "no-cache",
            range: `bytes=${offset}-${end}`,
          },
          signal: controller.signal,
        });
        if (response.status !== 206) {
          throw new Error(`GET ${url}: expected 206, got ${response.status}`);
        }
        const body = Buffer.from(await response.arrayBuffer());
        if (body.length !== end - offset + 1) {
          throw new Error(`short range ${body.length} != ${end - offset + 1}`);
        }
        fs.appendFileSync(partial, body);
        offset += body.length;
        if (offset - lastProgress >= PROGRESS_STEP || offset === expected) {
          console.log(
            `progress ${path.relative(modelRoot, destination)} `
            + `${offset}/${expected}`,
          );
          lastProgress = offset;
        }
        wrote = true;
        break;
      } catch (error) {
        if (attempt === 10) {
          throw error;
        }
        console.log(
          `retry ${attempt}/20 ${path.relative(modelRoot, destination)} `
          + `range=${offset}-${end}`,
        );
        await new Promise((resolve) => setTimeout(resolve, attempt * 2000));
      } finally {
        clearTimeout(timer);
      }
    }
    if (!wrote) {
      throw new Error(`failed to write range ${offset}-${end}`);
    }
  }

  const finalSize = await fileSize(partial);
  if (expected !== null && finalSize !== expected) {
    throw new Error(`${destination}: ${finalSize} != ${expected}`);
  }
  await rename(partial, destination);
  console.log(`downloaded ${path.relative(modelRoot, destination)} ${finalSize}`);
}

for (const checkpoint of CHECKPOINTS) {
  const manifestPath = path.join(modelRoot, checkpoint, "manifest.json");
  const manifestUrl = `${BASE_URL}/${checkpoint}/manifest.json`;
  await download(manifestUrl, manifestPath, await expectedSize(manifestUrl));
  const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));
  for (const fileName of manifest.files) {
    const destination = path.join(modelRoot, checkpoint, fileName);
    if ((await fileSize(destination)) !== null) {
      console.log(`skip ${path.relative(modelRoot, destination)}`);
      continue;
    }
    if (fileName === "model.safetensors" && !includeWeights) {
      throw new Error(`missing Marker model weight; rerun with --include-weights: ${destination}`);
    }
    const url = `${BASE_URL}/${checkpoint}/${fileName}`;
    const expected = await expectedSize(url);
    await download(url, destination, expected);
  }
}

console.log("datalab model cache complete");
